File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor xs
    | all isSpace xs = "Fine. Be that way!"
    | isShout        = "Whoa, chill out!"
    | isQuestion     = "Sure."
    | otherwise      = "Whatever."
  where isShout = (not . any isLower $ xs) && any isUpper xs
        isQuestion = last (filter (not . isSpace) xs) == '?'

File: cpp/bob/bob.cpp
#include "bob.h"

#include <boost/algorithm/string.hpp>
#include <boost/algorithm/string/predicate.hpp>

namespace bob {
    bool isQuestion(std::string const &address) {
	return boost::algorithm::ends_with(boost::trim_right_copy(address),
					   "?");
    }

    bool isEmpty(std::string const &address) {
	return boost::trim_right_copy(address).empty();
    }

    bool isShout(std::string const &address) {
	return
	    address.find_first_of("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
	    != std::string::npos
	    && address.find_first_of("abcdefghijklmnopqrstuvwxyz")
	    == std::string::npos;
    }

    std::string hey(std::string address) {
	return isShout(address) ? "Whoa, chill out!"
	    : isQuestion(address) ? "Sure."
	    : isEmpty(address) ? "Fine. Be that way!"
	    : "Whatever.";
    }
}
